docs(models): add reference and how-to for Platforms models. (#736)
canonical/craft-application: docs/howto/platforms.rst

Customize platforms in a project model
======================================

> Not all crafts can use the default model for platforms. When it's not possible, the

Author: IDK if we have uniform language for this

> application must perform this override in a prescribed manner:

Author: Changed - WDYT?

needed, the body of the class can be empty:

.. code-block:: python

    class MyPlatformsDict(GenericPlatformsDict[MyPlatform]):
        """A custom Platforms dict for MyCraft."""

If different keys can be considered shorthand values, you can replace
:attr:`~craft_application.models.GenericPlatformsDict._shorthand_keys` with an
appropriate set of shorthand keys:

.. code-block:: python

    class OnlyRiscv64ShorthandDict(GenericPlatformsDict[Platform]):
        """A custom PlatformsDict class that only allows "riscv64" as shorthand."""

        _shorthand_keys = [craft_platforms.DebianArchitecture.RISCV64]

If further customization is necessary, you may need to override
:meth:`~craft_application.models.GenericPlatformsDict.__get_pydantic_json_schema__`.
The
`Pydantic documentation <get_pydantic_json_schema_>`_
is a good place to start to understand this method.

Override ``platforms`` on the ``Project`` model
-----------------------------------------------

Once you have a non-generic child class of
:class:`~craft_application.models.GenericPlatformsDict`, override its type in your
project model:

.. code-block:: python

> 

Author: In the case of most actual apps, putting them both in the `project.py` file is fine. The only reason I moved it to `platforms.py` for craft-application itself is because we build a whole infrastructure for it. I guess I could caption them with `mycraft/models/project.py` for the suggested project structure?

> 

Reviewer: Sounds good to me.

    class MyProject(Project):
        """A Project class for MyCraft."""

        platforms: MyPlatformsDict

This will provide both your custom validation as necessary and a schema that implements
said validation.

Test your validation and schema
-------------------------------

> After building this, it is important to ensure that your schema and the actual project

Author: ```suggestion
After building a custom platforms model, it is important to ensure that your schema and the actual project
```

> After building this, it is important to ensure that your schema and the actual project

Reviewer: Perfect!

are treated the same way. This is necessary because the :ref:`ProjectService`
pre-processes the ``platforms`` key before validating the model. The best way to do
this is through a series of integration tests that compare the loading of valid files
to validation from the schema and vice versa.

> Craft-application does this with `a pair of integration tests

Reviewer 2: ```suggestion
Craft Application does this with `a pair of integration tests
```

Customize platforms in a project model
======================================

> Most applications can use the default model for platforms, if an application needs to

Reviewer: I think this is a run-on sentence.

Author: Good catch!

```suggestion
Most applications can use the default model for platforms. If an application needs to
```
